enhancements/machine-config/on-cluster-builds.md

```diff
@@ -0,0 +1,949 @@
+
+The first situation is image registry configuration for OpenShift
+installations without an internal image registry. We should add
+additional fields to the InstallConfig schema that allows cluster admins
+to specify the image registry to use and the push and pull secrets to
+use with it. While these could also be added as a Day 2 operation, I
+think it makes more sense to do it as a Day 1 operation here since one
+would theoretically know whether their OpenShift cluster will have an
+internal image registry.
+
+The second situation addresses use-cases where one wants to have custom
+OS content on a Day 1 basis. With that in mind, we can modify the
+InstallConfig schema to permit a Dockerfile to be provided on a per-pool
+basis. Where this would get a bit tricky is for multiarch clusters,
+however the current approach there is that bringing up a worker pool
+with a different architecture is currently a Day 2 operation. If / when
+that becomes a Day 1 operation, we’ll need to reconsider how this works.
+
+#### OpenShift CLI (oc)
+
+The oc command could gain the ability to export the MachineConfigs for a
+given MachineConfigPool as well as the bootc-native ConfigMaps that
+belong to it. Additionally, the oc command could provide a Podman
+incantation or Kube Pod Spec to ease the process of debugging the final
+container image and its configuration. It could look like this:
+
+`$ oc machineconfig export worker -o yaml`
+
+Or:
+
+`$ oc machineconfig export worker -o podman`
+
+Additionally, the oc command could gain additional abilities for cluster
+admins to control their on-cluster builds and the config / image rollout
+process:
+
+`$ oc machineconfig build worker`
+
+Or:
+
+`$ oc machineconfig rollout worker`
+
+Under the hood, this would make use of the annotations described in the
+MachineConfigPool API changes section to cause these actions to take
+place.
+
+### Risks and Mitigations
+
+What are the risks of this proposal and how do we mitigate. Think
+broadly. For example, consider both security and how this will impact
+the larger OKD ecosystem.
+
+How will security be reviewed and by whom?
+
+- For the custom pod builder, a review by someone who is knowledgeable
+  about container builders, RBAC, and SELinux would be helpful to ensure
+  that we can enable this behavior while avoiding the need for elevated
+  permissions / privileges.
+
+How will UX be reviewed and by whom?
+
+- TBD.
+
+### Drawbacks
+
+TBD.
+
+## Design Details
+
+### Open Questions
+
+- Since bootc can ultimately use ConfigMaps, does it make sense to
+  expose that functionality directly to cluster administrators by
+  allowing them to create / modify / delete ConfigMaps in the MCO
+  namespace to update their node configs? We could use a specific label
+  or annotation to signal to the MCO that it should consider these
+  ConfigMaps as it prepares the rendered MachineConfig.
+
+- What annotations will bootc support for ConfigMaps?
```

Comment:
I'd like to  try to track this stuff is best tracked in upstream discussions in bootc.  It's really up to us!  IOW *we* need to drive the design there so that it meets these use cases.

Reply:
Fine with me! I added some strawmans to this enhancement to get us started but I'm happy to add them someplace else.

Reply:
I put a slightly reduced set of my ConfigMap strawmans from this enhancement in https://github.com/containers/bootc/issues/22.